### PR TITLE
[Server] Backend support for Meshery Catalog

### PR DESCRIPTION
**Pull request by gr455**
**Description**
Adds catalog support to Meshery server backend
Exposes 4 new endpoints
`/api/pattern/clone/:id` and `/api/filter/clone/:id` to clone patterns and filters
`/api/pattern/catalog` and `/api/filter/catalog` to get catalog items

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.

### CODE REVIEW

**Comment by Revolyssup** [comment]
@gr455  Is this /fork endpoint added in cloud?

**Comment by gr455** [comment]
> @gr455  Is this /fork endpoint added in cloud?

Yes

**Comment by gr455** [comment]
// @MUzairS15

File: router/server.go [hunk 1]
```diff
@@ -150,6 +152,8 @@ func NewRouter(ctx context.Context, h models.HandlerInterface, port int, g http.
 		Methods("DELETE")
 	gMux.Handle("/api/filter/file/{id}", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.GetMesheryFilterFileHandler)))).
 		Methods("GET")
+	gMux.Handle("/api/filter/{id}/fork", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.ForkMesheryFilterHandler)))).
```
Comment (MUzairS15) [inline router/server.go:155]: @gr455 Can we make endpoint as `/api/filter/fork/{id}`. This resembles other endpoints.

**Comment by leecalcote** [comment]
> @gr455 Is this /fork endpoint added in cloud?

Which should be a `/clone` endpoint.

**Comment by gr455** [comment]
> > @gr455 Is this /fork endpoint added in cloud?
> 
> Which should be a `/clone` endpoint, if it isn't already.

Yes /clone has been added on cloud

**Comment by leecalcote** [comment]
Integration test failure for getting filters - https://github.com/meshery/meshery/runs/7937719167?check_suite_focus=true#step:6:113

**Comment by MUzairS15** [comment]
@Chadha93 Integration tests are failing as REST APIs are changed to GraphQL subscriptions.
It expects `/api/filter` request to happen when `filters` page is loaded, but now those are replaced by subscriptions, so `api/filter` request will never happen. Can you please update the tests with new logic??
I am not familiar with cypress.

**Comment by Chadha93** [comment]
> @Chadha93 Integration tests are failing as REST APIs are changed to GraphQL subscriptions. It expects `/api/filter` request to happen when `filters` page is loaded, but now those are replaced by subscriptions, so `api/filter` request will never happen. Can you please update the tests with new logic?? I am not familiar with cypress.

Sure, I'll look for this.

**Review by leecalcote — APPROVED** [review]

**Comment by MUzairS15** [comment]
@leecalcote @gr455 I just checked, on this branch MeshMap is breaking. Change required in MeshMap as well.
Working on it.

**Comment by MUzairS15** [comment]
Requires, succesive merging https://github.com/layer5labs/meshery-extensions/pull/1141

**Comment by Chadha93** [comment]
To fix this Filter spec integration failing test here due to the GraphQL, I think we need to refactor the whole spec file creating some GraphQL-specific utils and modifying queries accordingly. https://docs.cypress.io/guides/end-to-end-testing/working-with-graphql#Alias-multiple-queries-or-mutations. 

@Abhishek-kumar09 Need your suggestions here?

**Comment by leecalcote** [comment]
Merge conflicts have popped up.

File: ui/components/MesheryApplications.js [hunk 2]
```diff
@@ -248,21 +249,53 @@ function MesheryApplications({
...
+        search : search,
+        order : sortOrder
+      },
+      filterSelector : {
+        pageSize : pageSize.toString(),
+        page : page.toString(),
+        search : search,
+        order : sortOrder
+      }
+    });
```
Comment (YashKamboj) [inline ui/components/MesheryApplications.js:290]: @MUzairS15 I guess you wrote the same function here like the initAppsSubscription (I might be wrong)

Comment (MUzairS15) [reply ui/components/MesheryApplications.js:290]: Oops, you are right

Comment (YashKamboj) [reply ui/components/MesheryApplications.js:290]: Glad that i can help :handshake:

**Comment by gr455** [comment]
@MUzairS15 Please check the conflict resolution when you get time

**Comment by gr455** [comment]
@MUzairS15 Ready to merge this? Are the failing UI tests related?

**Comment by MUzairS15** [comment]
Tests are related, @Chadha93 Any updates??

**Comment by Chadha93** [comment]
> Tests are related, @Chadha93 Any updates?? (But can be handled separately??) // @Chadha93

Yes, The failing test is for Filters only, which needs to be changed according to the new graphQL req. I think we can be handled this separately and let this PR go. I will keep a note about the fix required for cypress tests.

**Comment by gr455** [comment]
> > Tests are related, @Chadha93 Any updates?? (But can be handled separately??) // @Chadha93
> 
> Yes, The failing test is for Filters only, which needs to be changed according to the new graphQL req. I think we can be handled this separately and let this PR go. I will keep a note about the fix required for cypress tests.

Okay then, @leecalcote can you merge this

**Review by Chadha93 — APPROVED** [review]
Merging this as approved.

**Comment by Chadha93** [comment]
Merging this so we can proceed with the catalog UI part and cypress separately.

**Pull request merged**